kernel/src/log_segment.rs

```diff
@@ -4,23 +4,183 @@
+        }
+
+        // Get the effective version from chosen files
+        let version_eff = sorted_commit_files
+            .last()
+            .or(checkpoint_parts.first())
+            .ok_or(Error::MissingVersion)? // TODO: A more descriptive error
+            .version;
+
+        Ok(LogSegment {
+            end_version: version_eff,
+            log_root,
+            commit_files: sorted_commit_files,
+            checkpoint_parts,
+        })
+    }
+
+    /// Constructs a [`LogSegment`] to be used for Snapshot. For a Snapshot at version `n`:
+    /// Its LogSegment is made of zero or one checkpoint, and all commits between the checkpoint up
+    /// to and including the end version `n`. Note that a checkpoint may be made of multiple
+    /// parts. All these parts will have the same checkpoint version.
+    ///
+    /// The options for constructing a LogSegment for Snapshot are as follows:
+    /// - `checkpoint_hint`: a `CheckpointMetadata` to start the log segment from (e.g. from reading the `last_checkpoint` file).
+    /// - `time_travel_version`: The version of the log that the Snapshot will be at.
+    #[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+    pub(crate) fn for_snapshot(
+        fs_client: &dyn FileSystemClient,
+        table_root: &Url,
+        checkpoint_hint: impl Into<Option<CheckpointMetadata>>,
+        time_travel_version: impl Into<Option<Version>>,
+    ) -> DeltaResult<Self> {
+        let time_travel_version = time_travel_version.into();
+        let log_root = table_root.join("_delta_log/").unwrap();
+
+        let (mut sorted_commit_files, checkpoint_parts) =
+            match (checkpoint_hint.into(), time_travel_version) {
+                (Some(cp), None) => {
+                    list_log_files_with_checkpoint(&cp, fs_client, &log_root, None)?
+                }
+                (Some(cp), Some(end_version)) if cp.version <= end_version => {
+                    list_log_files_with_checkpoint(&cp, fs_client, &log_root, Some(end_version))?
+                }
+                _ => list_log_files_with_version(fs_client, &log_root, None, time_travel_version)?,
+            };
+
+        // Commit file versions must be greater than the most recent checkpoint version if it exists
+        if let Some(checkpoint_file) = checkpoint_parts.first() {
+            sorted_commit_files.retain(|log_path| checkpoint_file.version < log_path.version);
+        }
+
+        let log_segment = LogSegment::try_new(sorted_commit_files, checkpoint_parts, log_root)?;
+
+        if let Some(time_travel_version) = time_travel_version {
+            require!(
+                log_segment.end_version == time_travel_version,
+                Error::generic(format!(
+                    "LogSegment end version {} not the same as the specified time_travel_version {}",
+                    log_segment.end_version, time_travel_version
+                ))
+            );
+        }
```

Comment:
It's not important that they're different. Initially I wasn't sure if passing an expected end version made sense since LogSegment wouldn't be constructed with it. But we might as well do the check in one place, so I went ahead and unified it.